pkg/controller/persistentpodstate/persistent_pod_state_event_handler.go

```diff
@@ -0,0 +1,280 @@
+	ref := metav1.GetControllerOf(pod)
+	// only statefulSet
+	if ref == nil || ref.Kind != KindSts.Kind {
+		return nil
+	}
+	ppsName := pod.Annotations[mutating.InjectedPersistentPodStateKey]
+	if ppsName != "" {
+		obj := &appsv1alpha1.PersistentPodState{}
+		if err := p.reader.Get(context.TODO(), client.ObjectKey{Namespace: pod.Namespace, Name: ppsName}, obj); err != nil {
+			klog.Errorf("fetch pod(%s/%s) PersistentPodState(%s) failed: %s", pod.Namespace, pod.Name, ppsName, err.Error())
+			return nil
+		}
+		return obj
+	}
+
+	return selectorPersistentPodState(p.reader, appsv1alpha1.TargetReference{
+		APIVersion: ref.APIVersion,
+		Kind:       ref.Kind,
+		Name:       ref.Name,
+	}, pod.Namespace)
+}
+
+func selectorPersistentPodState(reader client.Reader, ref appsv1alpha1.TargetReference, ns string) *appsv1alpha1.PersistentPodState {
+	ppsList := &appsv1alpha1.PersistentPodStateList{}
+	if err := reader.List(context.TODO(), ppsList, &client.ListOptions{Namespace: ns}, utilclient.DisableDeepCopy); err != nil {
+		klog.Errorf("List PersistentPodStateList failed: %s", err.Error())
+		return nil
+	}
+	for i := range ppsList.Items {
+		pps := &ppsList.Items[i]
+		if !pps.DeletionTimestamp.IsZero() {
+			continue
+		}
+		// belongs the same workload
+		if util.IsReferenceEqual(ref, pps.Spec.TargetReference) {
+			return pps
+		}
+	}
+	return nil
+}
+
+func isPodValidChanged(oldPod, newPod *corev1.Pod) bool {
+	if newPod.ResourceVersion == oldPod.ResourceVersion {
+		return false
+	}
+	// If the pod's deletion timestamp is set, reconcile
+	if oldPod.DeletionTimestamp.IsZero() && !newPod.DeletionTimestamp.IsZero() {
+		klog.V(3).Infof("pod(%s/%s) DeletionTimestamp changed, and reconcile", newPod.Namespace, newPod.Name)
+		return true
+	}
+
+	// when pod ready, then reconcile
+	if !podutil.IsPodReady(oldPod) && podutil.IsPodReady(newPod) {
+		klog.V(3).Infof("pod(%s/%s) from not ready to ready, and reconcile",
+			newPod.Namespace, newPod.Name)
+		return true
+	}
+	return false
+}
+
+var _ handler.EventHandler = &enqueueRequestForStatefulSet{}
+
+type enqueueRequestForStatefulSet struct {
+	reader client.Reader
+}
+
+func (p *enqueueRequestForStatefulSet) Create(evt event.CreateEvent, q workqueue.RateLimitingInterface) {
+	sts := evt.Object.(*appsv1.StatefulSet)
+	if sts.Annotations[appsv1alpha1.AnnotationAutoGeneratePersistentPodState] == "true" &&
+		(sts.Annotations[appsv1alpha1.AnnotationRequiredPersistentTopology] != "" ||
+			sts.Annotations[appsv1alpha1.AnnotationPreferredPersistentTopology] != "") {
+		enqueuePersistentPodStateRequest(q, KindSts.GroupVersion().String(), KindSts.Kind, sts.Namespace, sts.Name)
+	}
+}
+
+func (p *enqueueRequestForStatefulSet) Delete(evt event.DeleteEvent, q workqueue.RateLimitingInterface) {
+	sts := evt.Object.(*appsv1.StatefulSet)
+	if pps := selectorPersistentPodState(p.reader, appsv1alpha1.TargetReference{
```

Comment:
>  Get using naming convention (workload name and workload kind)
If the PPS is created by the user, it may not necessarily be workload.Name
> fetch pps using InjectedPersistentPodStateKey in the workload
workload annotations don't contains InjectedPersistentPodStateKey.

Most of all, delete statefulSet should be a very low-frequency operation, and it feels unnecessary to put very much effort into optimization.